### PR TITLE
Support custom entity deserializer in DeserializerFactory

### PR DESCRIPTION
**Pull request by Benestar**
To support more entity types we need to be able to pass a
DispatchableDeserializer in here that knows about all entity types, not
only then ones supported by Wikibase DataModel Serialization.

This patch uncouples the component from Wikibase DataModel Serialization.
It might be nice to get rid of that dependency completely.

Bug: [T126932](https://phabricator.wikimedia.org/T126932)

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
At present you can use this component without knowing that a lot of the stuff it does is basically just delegating to the other serialization component. This dependency is an implementation detail not visible from the outside. You are making it visible.

**Comment by Benestar** [comment]
The question is if it is desirable to have this invisble. What I have in mind is that we first collect all available deserializers for all entity types, then dispatch them and pass the resulting deserializer to this component. Do you have a suggestion how else we could do this?

**Comment by JeroenDeDauw** [comment]
It is certainly nicer to not have this visible. And it's not just the dependency on the other component your change here exposes, it is also the historical fact that we have two formats that still need to be supported.

I understand why you are making the change and do not object to it. I'm just pointing out that this is not a win encapsulation wise.

File: src/DeserializerFactory.php [hunk 1]
```diff
@@ -29,45 +29,33 @@ class DeserializerFactory {
...
 		);
 	}
 
 	/**
-	 * @since 1.1
-	 * @deprecated since 1.4 - use newStatementDeserializer instead
-	 *
-	 * @return Deserializer
-	 */
-	public function newClaimDeserializer() {
```
Comment (thiemowmde) [inline src/DeserializerFactory.php]: Split to a separate patch, please.

File: src/LegacyDeserializerFactory.php [hunk 2]
```diff
@@ -85,15 +85,6 @@ private function newSiteLinkListDeserializer() {
 	}
 
 	/**
-	 * @deprecated since 1.4 - use newStatementDeserializer instead
-	 *
-	 * @return Deserializer
-	 */
-	public function newClaimDeserializer() {
```
Comment (thiemowmde) [inline src/LegacyDeserializerFactory.php]: Split to a separate patch, please.

File: src/DeserializerFactory.php [hunk 3]
```diff
@@ -29,45 +29,33 @@ class DeserializerFactory {
...
-		$this->currentFactory = new CurrentDeserializerFactory( $dataValueDeserializer, $idParser );
 	}
 
 	/**
+	 * @param DispatchableDeserializer $currentEntityDeserializer
+	 *
 	 * @return Deserializer
 	 */
-	public function newEntityDeserializer() {
+	public function newEntityDeserializer( DispatchableDeserializer $currentEntityDeserializer ) {
```
Comment (thiemowmde) [inline src/DeserializerFactory.php]: I'm not objecting this change, at least not it's intention. But this method signature is now really weird and confusing. Why do you have to pass an entity serializer to a method called `newEntitySerializer`? Sounds like somebody suggests to build endless loops. Can we rename this stuff somehow so it becomes more obvious what all this is about?

File: src/DeserializerFactory.php [hunk 4]
```diff
@@ -29,45 +29,33 @@ class DeserializerFactory {
...
-
-	/**
 	 * @since 1.4
 	 *
+	 * @param DispatchableDeserializer $currentStatementDeserializer
+	 *
 	 * @return Deserializer
 	 */
-	public function newStatementDeserializer() {
+	public function newStatementDeserializer( DispatchableDeserializer $currentStatementDeserializer ) {
```
Comment (thiemowmde) [inline src/DeserializerFactory.php]: How is this related to the patch's intention, to allow other entity types?

File: src/DeserializerFactory.php [hunk 5]
```diff
@@ -29,45 +29,33 @@ class DeserializerFactory {
...
-	 * @deprecated since 1.4 - use newStatementDeserializer instead
-	 *
-	 * @return Deserializer
-	 */
-	public function newClaimDeserializer() {
-		return $this->newStatementDeserializer();
-	}
-
-	/**
 	 * @since 1.4
```
Comment (thiemowmde) [inline src/DeserializerFactory.php]: The `@since` tags of the incompatible methods should change.

File: src/DeserializerFactory.php (hunk 3 above)
Comment (Benestar) [inline src/DeserializerFactory.php]: Maybe we should pass the deserializer in the constructor? That would perhaps make more sense..

**Comment by Benestar** [comment]
I changed this patch and now it contains no breaking changes anymore but only adds some new functionality to provide a custom entity deserializers instead of using the factory.

File: tests/integration/DeserializerFactoryTest.php [hunk 6]
```diff
@@ -21,9 +23,25 @@ protected function setUp() {
...
 
-	public function testGetEntityDeserializerReturnsDeserializer() {
+	public function testNewEntityDeserializerReturnsDeserializer() {
 		$deserializer = $this->factory->newEntityDeserializer();
 		$this->assertInstanceOf( 'Deserializers\Deserializer', $deserializer );
 	}
 
+	public function testNewStatementDeserializerReturnsDeserializer() {
+		$deserializer = $this->factory->newStatementDeserializer();
+		$this->assertInstanceOf( 'Deserializers\Deserializer', $deserializer );
```
Comment (JeroenDeDauw) [inline tests/integration/DeserializerFactoryTest.php]: bumping to PHP to min 5.5 avoids having to use strings for this, as per example at the bottom of https://www.entropywins.wtf/blog/2016/01/30/missing-in-php7-function-references/

File: src/DeserializerFactory.php [hunk 7]
```diff
@@ -45,7 +56,7 @@ public function __construct( Deserializer $dataValueDeserializer, EntityIdParser
 	public function newEntityDeserializer() {
 		return new EntityDeserializer(
 			$this->legacyFactory->newEntityDeserializer(),
-			$this->currentFactory->newEntityDeserializer()
+			$this->currentEntityDeserializer ?: $this->currentFactory->newEntityDeserializer()
```
Comment (JeroenDeDauw) [inline src/DeserializerFactory.php]: I think it'd be good to describe this behaviour for the constructor parameter. Else it's not clear that the deserializer you provide gets used instead of some default.

**Comment by JeroenDeDauw** [comment]
Generally looks good

File: tests/integration/DeserializerFactoryTest.php (hunk 6 above)
Comment (Benestar) [inline tests/integration/DeserializerFactoryTest.php]: I like the new PHP 5.5 feature but we should do this once for all tests to stay consistent.

File: src/DeserializerFactory.php (hunk 7 above)
Comment (Benestar) [inline src/DeserializerFactory.php]: Added a comment to the constructor

File: src/DeserializerFactory.php [hunk 8]
```diff
@@ -34,9 +35,25 @@ class DeserializerFactory {
...
+	 * @param DispatchableDeserializer|null $currentEntityDeserializer used instead of constructing
+	 *        a new current Deserializer for entities using a current DeserializerFactory.
+	 */
+	public function __construct(
+		Deserializer $dataValueDeserializer,
+		EntityIdParser $idParser,
+		DispatchableDeserializer $currentEntityDeserializer = null
+	) {
 		$this->legacyFactory = new LegacyDeserializerFactory( $dataValueDeserializer, $idParser );
 		$this->currentFactory = new CurrentDeserializerFactory( $dataValueDeserializer, $idParser );
```
Comment (brightbyte) [inline src/DeserializerFactory.php]: If $currentEntityDeserializer is given, $this->currentFactory should never be used, right? So perhaps set it to null.

Alternatively, we could have a trivial DeserializerFactory that just always returns the same serializer. Then we could just do   $this->currentFactory = new SingleDeserializerFactory( $currentEntityDeserializer ). 

Or we could have two implementations of DeserializerFactory - one that takes a factory for the current serializers, and one that takes a serializer. 

It all seems a bit silly. Just provide a way to get the legacy serializers, and let the caller sort it out.

**Comment by brightbyte** [comment]
No strong objections. We can do it this way, but it's not exactly pretty. Let's not get stuck on this. If we need it, let's merge it.

File: src/DeserializerFactory.php (hunk 8 above)
Comment (Benestar) [inline src/DeserializerFactory.php]: That is not correct, since `$this->currentFactory` is also used in the `newStatementDeserializer` method below.

Comment (brightbyte) [inline src/DeserializerFactory.php]: Huh, right. Looks like newStatementDeserializer should die though: it's only used (via WikibaseRepo::getInternalStatementDeserializer and EntityChange::getStatementDeserializer) to deserialize change rows that contain old style serialization of statements. We needed that fro a transition period for about a week a few years ago. I think it can die now (along with WikibaseRepo::getInternalStatementDeserializer and EntityChange::getStatementDeserializer).

Comment (brightbyte) [inline src/DeserializerFactory.php]: Filed as https://phabricator.wikimedia.org/T128618
